content/blog/2023-12-01-task-lifecycle.markdown

```diff
@@ -0,0 +1,470 @@
+	 * CPUs for init to the non isolated CPUs.
+	 */
+	rcu_read_lock();
+	tsk = find_task_by_pid_ns(pid, &init_pid_ns);
+	tsk->flags |= PF_NO_SETAFFINITY;
+	set_cpus_allowed_ptr(tsk, cpumask_of(smp_processor_id()));
+	rcu_read_unlock();
+
+	numa_default_policy();
+	pid = kernel_thread(kthreadd, NULL, NULL, CLONE_FS | CLONE_FILES);
+	rcu_read_lock();
+	kthreadd_task = find_task_by_pid_ns(pid, &init_pid_ns);
+	rcu_read_unlock();
+
+	/*
+	 * Enable might_sleep() and smp_processor_id() checks.
+	 * They cannot be enabled earlier because with CONFIG_PREEMPTION=y
+	 * kernel_thread() would trigger might_sleep() splats. With
+	 * CONFIG_PREEMPT_VOLUNTARY=y the init task might have scheduled
+	 * already, but it's stuck on the kthreadd_done completion.
+	 */
+	system_state = SYSTEM_SCHEDULING;
+
+	complete(&kthreadd_done);
+
+	/*
+	 * The boot idle thread must execute schedule()
+	 * at least once to get things moving:
+	 */
+	schedule_preempt_disabled();
+	/* Call into cpu_idle with preempt disabled */
+	cpu_startup_entry(CPUHP_ONLINE);
+}
+
+```
+
+After this, it basically just enters the scheduler. We're now almost fully task
+driven, and our journey picks back up inside of the init task.
+
+> Hint: you can see where the `init_task` vs the init task is used in the
+> console. The `init_task` has PID 0, so appears as `T0` in the log line
+> metadata. The init task has PID 1, so appears as `T1`.
+
+
+## The init task
+
+When we created the init task, we set the entry point to be the `kernel_init()`
+function. Execution simply begins from here once we get scheduled. The very
+first thing we do is wait for the kthreadd task to be up and running: bad things
+would happen if we tried to use kernel services that rely on kthreads otherwise.
+
+The wait mechanism itself also interacts with the scheduler a lot: starting with
+a common `struct completion` object, the waiting task registers itself as
+awaiting the object to complete. Specifically, it adds its task handle to a
+queue on the completion object. It then loops calling `schedule()`, which yields
+itself to the scheduler, until the completion object is flagged as done.
+Somewhere else, such as another task, the completion object will eventually be
+marked as completed. As part of this the task marking the completion tries to
+wake up a task that might have registered itself as waiting earlier.
+
+In the kthreadd case, the init task waits on a completion object that
+`kernel_init()` marks completed after creating kthreadd. We could technically
+avoid this synchronization altogether just by creating kthreadd first, but then
+the init task wouldn't have PID 1.
+
+The rest of the init task wraps up the initialisation stage as a whole. Mostly
+it moves the system into the 'running' state after freeing any memory marked as
+for initialisation only (set by `__init` annotations).
+
+Once fully initialised and running, the init tasks attempts to execute the
+userspace init program.
```

Comment:
May help to itemise the possibilities for what an init program could be (systemd, Sys-V init, or a custom script etc), just to prime readers for understanding the user/kernel-space dichotomy.